barretenberg/cpp/src/barretenberg/smt_verification/circuit/ultra_circuit.cpp

```diff
@@ -401,6 +450,298 @@ void UltraCircuit::handle_range_constraints()
+{
+    if (this->public_inps.contains(ram_index_idx) || this->rom_ram_relaxed) {
+        STerm index = this->symbolic_vars[ram_index_idx];
+        index == this->variables[ram_index_idx];
+    }
+
+    SymArray<STerm, STerm>& ram_table = this->cached_ram_tables[ram_array_idx];
+    STerm index = this->symbolic_vars[ram_index_idx];
+    STerm value_entry = this->symbolic_vars[read_from_value_idx];
+
+    ram_table.put(index, value_entry);
+}
+
+/**
+ * @brief Adds all the ROM related constraints into the solver.
+ *
+ */
+void UltraCircuit::handle_rom_tables()
+{
+    static constexpr uint32_t UNINITIALIZED_MEMORY_RECORD = UINT32_MAX;
+
+    STerm idx_ex = this->symbolic_vars[this->variable_names_inverse["zero"]];
+    STuple entry_ex({ idx_ex, idx_ex });
+
+    cvc5::Sort ind_sort = idx_ex.term.getSort();
+    TermType ind_type = idx_ex.type;
+    cvc5::Sort entry_sort = entry_ex.term.getSort();
+    TermType entry_type = entry_ex.type;
+
+    for (uint32_t i = 0; i < this->rom_records.size(); i++) {
+        SymArray<STerm, STuple> rom_table(
+            ind_sort, ind_type, entry_sort, entry_type, this->solver, "ROM_TABLE#" + std::to_string(i));
+        // Fill the ROM table
+        for (size_t j = 0; j < this->rom_states[i].size(); j++) {
+            STerm idx(static_cast<bb::fr>(j), this->solver, ind_type);
+            if (this->rom_states[i][j][0] == UNINITIALIZED_MEMORY_RECORD) {
+                continue;
+            }
+
+            STerm value1 = this->symbolic_vars[this->rom_states[i][j][0]];
+            STerm value2 = this->symbolic_vars[this->rom_states[i][j][1]];
+            rom_table.put(idx, STuple({ value1, value2 }));
+        }
+        this->cached_rom_tables.insert({ i, rom_table });
+
+        // process all the reads
+        for (auto rom_record : this->rom_records[i]) {
+            uint32_t index_witness = rom_record[0];
+            uint32_t value1_witness = rom_record[1];
+            uint32_t value2_witness = rom_record[2];
+            this->rom_table_read(i, index_witness, value1_witness, value2_witness);
+        }
+    }
+}
+
+/**
+ * @brief Adds all the RAM related constraints into the solver.
+ *
+ */
+void UltraCircuit::handle_ram_tables()
+{
+    STerm idx_ex = this->symbolic_vars[this->variable_names_inverse["zero"]];
+    STuple entry_ex({ idx_ex, idx_ex });
+
+    cvc5::Sort sort = idx_ex.term.getSort();
+    TermType type = idx_ex.type;
+
+    for (uint32_t i = 0; i < this->ram_records.size(); i++) {
+        SymArray<STerm, STerm> ram_table(sort, type, sort, type, this->solver, "RAM_TABLE#" + std::to_string(i));
+        this->cached_ram_tables.insert({ i, ram_table });
+
+        // process all the reads and writes
+        for (auto ram_record : this->ram_records[i]) {
+            uint32_t index_witness = ram_record[0];
+            uint32_t value_witness = ram_record[1];
+            // uint32_t timestamp_witness = ram_record[2];
```

Comment:
I'd ensure that the indices are sorted. Technically there shouldn't be an issue because of how we construct, but if you are just looking at the proving system, it's possible